### PR TITLE
Better code display for 1 instruction inf loop #180

### PR DESCRIPTION
**Pull request by disconnect3d**
Removes all duplicate instructions from the tail end of the disassebled instructions list.

### CODE REVIEW

File: pwndbg/commands/nearpc.py [hunk 1]
```diff
@@ -80,7 +80,23 @@ def nearpc(pc=None, lines=None, to_string=False, emulate=False):
     #             pc_to_linenos[line.pc].append(line.line)
     instructions = pwndbg.disasm.near(pc, lines, emulate=emulate)
```
Comment (zachriggle) [inline pwndbg/commands/nearpc.py]: I would much rather this logic be implemented in `pwndbg.disasm.near`, rather than in the command itself.

**Comment by zachriggle** [comment]
I was playing around with this a bit, and have one suggestion and one problem.

Suggestion is that I think this is a bit more concise:

```py
    # Remove infinite loops repeating themselves
    while insns and len(insns) > 1 and insns[-1] == insns[-2]:
        del insns[-1]
```

The problem with doing the unrolling this way is that we're thinking of infinite loops, but the forward-emulation for ROP gadgets will break for a RET sled under this model.  It'll just look like a single RET, and then nothing more.

I think a fair way to do this would be to let the instruction repeat *once* (i.e. the last two instructions are the same), so that we can at least see the "next" instruction / gadget.

**Comment by disconnect3d** [comment]
Do you mean something like this by a RET sled?:
```asm
[bits 32]

global _start
_start:

call test1
ret
test1:
  call test2
  ret
test2:
  call test3
  ret
test3:
  call test4
  ret
test4:
  call test5
  ret
test5:
    ret
```

Launching:
```
nasm -f elf test.asm && ld -m elf_i386 test.o -o test && gdb -q ./test
```

This produces the same output with and without your changes.
Also the output is... pretty weird for me, see https://asciinema.org/a/01iwnxnmxxd3jn2kjhfdb5jiw

**Comment by zachriggle** [comment]
No, in your case all of the `ret` instructions are at different addresses.

The reason that the output is weird is because of the "previous instruction" tracking.  At `test4`, we call `test5`.  `test5` is just a `ret`, and the Unicorn emulation lets us know that the next instruction is the `ret` inside `test4`, and so on.  The **forward** instructions are all correct.  The **backward** instructions get a bit confused since we're going `test5 ret` --> `test4 ret`, so we save `test5` as the instruction which "preceded" `test4 ret`.  So when we unfold backward, things get weird.

The "ret sled" I'm referring to would be something like this (haven't tested it, but you should get the idea)

```
## "Terminate" our ROP stack with an CC
    call A
    int3
A:

## Push the "ret" address onto the stack a whole bunch
    mov ecx, 0x100
THELOOP:
    loop push_retaddr
    ret

## Push the "ret" address via "call"
push_ret_sled_address:
    call THELOOP

## Actual "ret sled" consists of this address on the stack a whole bunch,
## which Unicorn will emulate through so that we know the "next" instruciton
## is still this one
    ret
```

**Comment by zachriggle** [comment]
screen shot 2017-04-05 at 2 48 43 pm

**Comment by disconnect3d** [comment]
Closed - we've added an alternate implementation: https://github.com/pwndbg/pwndbg/pull/205.

**Pull request closed without merging**